keps/sig-scheduling/2779-support-custom-queue-for-the-scheduler/README.md

```diff
@@ -0,0 +1,423 @@
+and get some functions that are not supported now.
+
+## Motivation
+The current queue (PriorityQueue in pkg\scheduler\internal\queue\scheduling_queue.go) 
+for pod selection in the scheduler is good enough for most cases, while it cannot handle 
+all the requests even with a custom scheduler based on the kubernetes scheduler plugins.
+
+One example:
+There are many pods from 2 users (userA and userB), and the target is to ensure resource
+usage (E.g. userA gets x CPUs) ratio between userA and userB is 1:1. Sorting the pods by
+the resource usage in Less function cannot work, because the resource usage for 
+userA/userB will be updated dynamically once a pod is selected and bound, while the 
+algorithm (heapsort) that used by the current queue to get the next pods depends on static
+data.
+
+Another example:
+The interval of function flushUnschedulableQLeftover is hard coded to 30 seconds, user
+cannot change it.
+
+These issues can be resolved if user can implement the queue by themselves, then user
+can get what they want easily, and code maintainer needn't handle various requests.
+
+### Goals
+
+Support custom queue of the scheduler.
+
+### Non-Goals
+
+Support custom rules for pod selection in the current queue of scheduler
+
+## Proposal
+
+Provide an interface like current scheduler plugins, users can provide a
+custom queue of scheduler and register it, then the kubernetes
+scheduler will use this custom queue for pod selection.
+
+This is an extension of the current scheduler plugins, users can control
+more details of the scheuler with this enhancement.
+
+Pros:  
+Users can get full control of the pod selection logic in the queue, they
+can select pods with custom logic.
+
+The scheduler plugins design is enhanced, the custom queue and the extension
+points can work together to meet more requests.
+
+Cons:  
+Users need to understand the details of the current queue which has 13
+functions now, it is not easy to implement a custom queue.
+
+### Risks and Mitigations
+
+If custom queue is not enabled or provided, the current queue will work as
+before, no other users will be impacted.
+
+## Design Details
+
+The codes will be updated as following:
+
+1. User implements a custom queue with the interface SchedulingQueue,
+the name is myQueue, and registers it
+
+  ```
+	command := app.NewSchedulerCommand(
+		app.WithPlugin(coscheduling.Name, coscheduling.New),
+                        …...
+	+	app.WithCustomQueue(myQueue.New),
+	)
+  ```
+
+2. The registered custom queue will be recorded by updated Registry struct.
+
+  ```
+  type Registry struct {
+    Pf          map[string]PluginFactory
+    CustomQueue schedulingQueue.SchedulingQueue
+  }
+  ```
+
+3. In function (c *Configurator) create(), pass the custom queue to the scheduler
+if it exists.
+  ```
+  if c.registry.CustomQueue != nil {
+      podQueue = c.registry.CustomQueue
+  } else {
+      podQueue = schedulingQueue.NewSchedulingQueue()
+  }
+   ```
+
```

Comment:
Can you also add here to explain how it explains with other internal scheduler structures like Cache, Assumed state, NominatedPods, etc. Any other interfaces are needed? and how. What's the risk here?

Reply:
A new interface Init will be aded to handle the initialization parameters from the k8s framework, it is used to replace the old function NewPriorityQueue, the logic inside them are same.
Packages queue and heap will be exposed to end users, then users can access/use them and the mentioned internal scheduler structures. 


```
4. Basic inputs for queue's initialization will not be changed, a new interface `Init` is added in
SchedulingQueue, users must implement it.
-func NewPriorityQueue(
+func (p *PriorityQueue) Init(
 	lessFn framework.LessFunc,
 	informerFactory informers.SharedInformerFactory,
 	opts ...Option,
)
5. For the scheduler internal packages under ([internal] (https://github.com/kubernetes/kubernetes/tree/b6c75bee15e150628fcc240ab32dba6190d254e4/pkg/scheduler/internal)), queue and heap will be moved out of this folder so that users can access and use the structures inside them, cache and parallelize will not be touched.
```

Reply:
> Packages queue and heap will be exposed to end users, then users can access/use them and the mentioned internal scheduler structures.

I don't think I like this idea. The internal structure should be restricted for internal usage only.

Can you list the blockers (due to missing plumbing mechanics) that a pure out-of-tree implementation won't work properly?

Also, I'm also thinking about how `Option` can be more general and thus more usable for an out-of-tree queue implementation. Right now the internal `priorityQueueOptions` are fixed data structures, and hence not extensible.

Reply:
OK, let's keep queue and heap in the original location.  Removed `5` from the document.

The current blockers in my mind are `SchedulingQueue` and `PriorityQueueOptions` (maybe there will be more, I will write some codes to verify it), and I'm planning to move them to `pkg\scheduler\framework\interface.go` so that the users can access them, this is a right loaction for you?

Reply:
close this thread as there is same thread below.